Entity literal preferred over path in parsing when entity name and attribute names conflict

### Description
Hi, I recently discovered Blaze Persistence and so far has been great at solving some issues I was having. I have arrived at a particular issue that I cant seem to find the solution for.
I am trying to use nested view for ManyToOne mapping but gets an error, this is not a bug report but more of a request for help.


### Entities
_Some fields removed for simplicity_

#### Subject
```
@Audited
@Entity(name = "curriculum_subject")
public class SubjectImpl implements Subject {

    private String uid;

    @JsonDeserialize(as = SchoolImpl.class)
    @JsonFilter(JsonFilterConstants.UID_ONLY)
    private School school;

    @Id
    @Column(name = "id")
    @GenericGenerator(name  = "uid64", strategy = "com.-.data.util.UID64Generator")
    @GeneratedValue(generator = "uid64")
    public String getUid() {
        return uid;
    }

    public void setUid(String uid) {
        this.uid = uid;
    }

    @ManyToOne(targetEntity = SchoolImpl.class, optional = false, fetch = FetchType.LAZY)
    public School getSchool() {
        return school;
    }

    public void setSchool(School school) {
        this.school = school;
    }
}
```

#### School
```
@Audited
@Entity(name = "school")
@NoArgsConstructor
public class SchoolImpl implements School {

    private String uid;

    private String name;

    @Id
    @Column(name = "id")
    @GenericGenerator(name  = "uid64", strategy = "com.-.data.util.UID64Generator")
    @GeneratedValue(generator = "uid64")
    public String getUid() {
        return uid;
    }

    public void setUid(String uid) {
        this.uid = uid;
    }

    @Column(name = "school_name")
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
}
```



### Views
_These are the complete views, nothing removed_

#### Subject
```
@EntityView(SubjectImpl.class)
public interface SubjectView extends DomainView {
    String getName();
    String getShortName();
    SchoolView getSchool();
}
```

#### School
```
@EntityView(SchoolImpl.class)
public interface SchoolView extends DomainView {
    String getName();
}
```

#### Domain
```
public interface DomainView {
    @IdMapping
    String getUid();
}
```

### Stacktrace
Whenever I try to build I get the following error
```
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'createEntityViewManager' defined in class path resource [com/-/main/configuration/BlazePersistenceConfig.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.blazebit.persistence.view.EntityViewManager]: Factory method 'createEntityViewManager' threw exception; nested exception is java.lang.IllegalArgumentException: There are error(s) in entity views!
The resolved possible types [javax.persistence.metamodel.EntityType] are not assignable to the given expression type 'com.-.domain.modules.school.entity.impl.SchoolImpl' of the mapping expression declared by the attribute school[com.-.domain.modules.curriculum.view.SubjectView.getSchool]!
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:656)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:636)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1338)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1177)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:557)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:517)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:323)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:226)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:321)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:276)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1306)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1226)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:640)
	... 95 more
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.blazebit.persistence.view.EntityViewManager]: Factory method 'createEntityViewManager' threw exception; nested exception is java.lang.IllegalArgumentException: There are error(s) in entity views!
The resolved possible types [javax.persistence.metamodel.EntityType] are not assignable to the given expression type 'com.-.domain.modules.school.entity.impl.SchoolImpl' of the mapping expression declared by the attribute school[com.-.domain.modules.curriculum.view.SubjectView.getSchool]!
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:651)
	... 108 more
Caused by: java.lang.IllegalArgumentException: There are error(s) in entity views!
The resolved possible types [javax.persistence.metamodel.EntityType] are not assignable to the given expression type 'com.-.domain.modules.school.entity.impl.SchoolImpl' of the mapping expression declared by the attribute school[com.-.domain.modules.curriculum.view.SubjectView.getSchool]!
	at com.blazebit.persistence.view.impl.EntityViewManagerImpl.<init>(EntityViewManagerImpl.java:275)
	at com.blazebit.persistence.view.impl.EntityViewConfigurationImpl.createEntityViewManager(EntityViewConfigurationImpl.java:206)
	at com.-.main.configuration.BlazePersistenceConfig.createEntityViewManager(BlazePersistenceConfig.java:35)
	at com.-.main.configuration.BlazePersistenceConfig$$EnhancerBySpringCGLIB$$76fc933b.CGLIB$createEntityViewManager$0(<generated>)
	at com.-.main.configuration.BlazePersistenceConfig$$EnhancerBySpringCGLIB$$76fc933b$$FastClassBySpringCGLIB$$39a92948.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331)
	at com.-.main.configuration.BlazePersistenceConfig$$EnhancerBySpringCGLIB$$76fc933b.createEntityViewManager(<generated>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
	... 109 more
```

With the above error while building application I can see the following generated classes
#### SubjectView_
```
@Generated(value = "com.blazebit.persistence.view.processor.EntityViewAnnotationProcessor")
@StaticMetamodel(SubjectView.class)
public abstract class SubjectView_ {
    public static volatile MethodSingularAttribute<SubjectView, String> name;
    public static volatile SchoolViewRelation<SubjectView, MethodSingularAttribute<SubjectView, SchoolView>> school;
    public static volatile MethodSingularAttribute<SubjectView, String> shortName;
    public static volatile MethodSingularAttribute<SubjectView, String> uid;

    public static final String NAME = "name";
    public static final String SCHOOL = "school";
    public static final String SHORT_NAME = "shortName";
    public static final String UID = "uid";
    public static final String SCHOOLNAME = "school.name";
    public static final String SCHOOLUID = "school.uid";

    public static EntityViewSetting<SubjectView, CriteriaBuilder<SubjectView>> createSettingInit() {
        return EntityViewSetting.create(SubjectView.class, "init");
    }
    public static EntityViewSetting<SubjectView, PaginatedCriteriaBuilder<SubjectView>> createPaginatedSettingInit(int firstResult, int maxResults) {
        return EntityViewSetting.create(SubjectView.class, firstResult, maxResults, "init");
    }
}
```

#### SubjectViewImpl
```
@Generated(value = "com.blazebit.persistence.view.processor.EntityViewAnnotationProcessor")
@StaticImplementation(SubjectView.class)
public class SubjectViewImpl implements SubjectView, EntityViewProxy {

    public static volatile EntityViewManager ENTITY_VIEW_MANAGER;
    public static final SerializableEntityViewManager SERIALIZABLE_ENTITY_VIEW_MANAGER = new SerializableEntityViewManager(SubjectViewImpl.class, ENTITY_VIEW_MANAGER);

    private final String name;
    private final SchoolView school;
    private final String shortName;
    private final String uid;

    public SubjectViewImpl(SubjectViewImpl noop, Map<String, Object> optionalParameters) {
        this.name = null;
        this.school = null;
        this.shortName = null;
        this.uid = null;
    }

    public SubjectViewImpl(        String uid) {
        this.$$_kind = (byte) 1;
        this.name = null;
        this.school = null;
        this.shortName = null;
        this.uid = uid;
    }

    public SubjectViewImpl(
        String uid,
        String name,
        SchoolView school,
        String shortName
    ) {
        super();
        this.name = name;
        this.school = school;
        this.shortName = shortName;
        this.uid = uid;
    }

    public SubjectViewImpl(SubjectViewImpl noop, int offset, Object[] tuple) {
        super();
        this.name = (String) tuple[offset + 1];
        this.school = (SchoolView) tuple[offset + 2];
        this.shortName = (String) tuple[offset + 3];
        this.uid = (String) tuple[offset + 0];
    }

    public SubjectViewImpl(SubjectViewImpl noop, int offset, int[] assignment, Object[] tuple) {
        super();
        this.name = (String) tuple[offset + assignment[1]];
        this.school = (SchoolView) tuple[offset + assignment[2]];
        this.shortName = (String) tuple[offset + assignment[3]];
        this.uid = (String) tuple[offset + assignment[0]];
    }


    @Override
    public String getName() {
        return name;
    }
    @Override
    public SchoolView getSchool() {
        return school;
    }
    @Override
    public String getShortName() {
        return shortName;
    }
    @Override
    public String getUid() {
        return uid;
    }
...
```

### Environment
Version:  1.5.1          
JPA-Provider:   Hibernate 5.4.15.Final
DBMS:  Postgres 11             
Application Server: Java EE with Spring Data

If I remove `getSchool()` from the SubjectView the error goes away. Any idea what I am doing wrong or missing. I am still going through the documentations and appreciate any pointers that could help in my troubles. Hopefully the information provided is enough, thank you.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -34,6 +34,7 @@ None yet
 * Fix issues with de-serializing of singular entity view attributes that use a collection type
 * Add Cockroach DBMS detection and a `DbmsDialect` implementation
 * Fix various issues with type resolving paths that match entity names
+* Fix issues with remove view by id when element type of cascaded collection uses embedded id
 
 ### Backwards-incompatible changes
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -33,6 +33,7 @@ None yet
 * Workaround Hibernate proxy field access bug for non-pk relations
 * Fix issues with de-serializing of singular entity view attributes that use a collection type
 * Add Cockroach DBMS detection and a `DbmsDialect` implementation
+* Fix various issues with type resolving paths that match entity names
 
 ### Backwards-incompatible changes
 

--- a/core/impl/src/main/java/com/blazebit/persistence/impl/JoinManager.java
+++ b/core/impl/src/main/java/com/blazebit/persistence/impl/JoinManager.java
@@ -2866,7 +2866,7 @@ public class JoinManager extends AbstractManager<ExpressionModifier> {
                                 Type<?> type = attributeHolder.getAttributeType();
                                 result = new JoinResult(currentResult.baseNode, Arrays.asList(elementString), type, -1, -1);
                             } else if (metamodel.getManagedType(ExtendedManagedType.class, current.getManagedType()).getAttributes().get(associationName) != null) {
-                                Expression resultExpr = expressionFactory.createSimpleExpression(associationName, false);
+                                Expression resultExpr = new PathExpression(new PropertyExpression(associationName));
                                 AttributeHolder attributeHolder = JpaUtils.getAttributeForJoining(
                                         metamodel,
                                         current.getNodeType(),
@@ -3475,13 +3475,11 @@ public class JoinManager extends AbstractManager<ExpressionModifier> {
         if (resultFields.isEmpty()) {
             return new JoinResult(current, null, current == null ? null : current.getNodeType(), singleValuedAssociationNameStartIndex, singleValuedAssociationNameEndIndex);
         } else {
-            StringBuilder sb = new StringBuilder();
-            sb.append(resultFields.get(0));
-            for (int i = 1; i < resultFields.size(); i++) {
-                sb.append('.');
-                sb.append(resultFields.get(i));
+            List<PathElementExpression> pathElementExpressions = new ArrayList<>(resultFields.size());
+            for (int i = 0; i < resultFields.size(); i++) {
+                pathElementExpressions.add(new PropertyExpression(resultFields.get(i)));
             }
-            Expression expression = expressionFactory.createSimpleExpression(sb.toString(), false);
+            Expression expression = new PathExpression(pathElementExpressions);
             Type<?> type = JpaUtils.getAttributeForJoining(metamodel, current.getNodeType(), expression, null).getAttributeType();
             return new JoinResult(current, resultFields, type, singleValuedAssociationNameStartIndex, singleValuedAssociationNameEndIndex);
         }

--- a/core/parser/src/main/java/com/blazebit/persistence/parser/expression/AbstractExpressionFactory.java
+++ b/core/parser/src/main/java/com/blazebit/persistence/parser/expression/AbstractExpressionFactory.java
@@ -206,6 +206,14 @@ public abstract class AbstractExpressionFactory extends AbstractExpressionFactor
             List<PathElementExpression> pathElements = new ArrayList<>(1);
             pathElements.add((PathElementExpression) expr);
             return new PathExpression(pathElements);
+        } else if (expr instanceof EntityLiteral) {
+            // Special case where the path matches an entity name
+            String[] parts = ((EntityLiteral) expr).getOriginalExpression().split("\\.");
+            List<PathElementExpression> pathElements = new ArrayList<>(parts.length);
+            for (String part : parts) {
+                pathElements.add(new PropertyExpression(part));
+            }
+            return new PathExpression(pathElements);
         } else if (expr instanceof FunctionExpression) {
             return expr;
         } else {

--- a/core/testsuite/src/main/java/com/blazebit/persistence/testsuite/entity/DocumentHolder.java
+++ b/core/testsuite/src/main/java/com/blazebit/persistence/testsuite/entity/DocumentHolder.java
@@ -1,0 +1,62 @@
+/*
+ * Copyright 2014 - 2021 Blazebit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.blazebit.persistence.testsuite.entity;
+
+import javax.persistence.Entity;
+import javax.persistence.FetchType;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.OneToOne;
+import java.io.Serializable;
+
+/**
+ *
+ * @author Christian Beikov
+ * @since 1.6.0
+ */
+@Entity
+public class DocumentHolder implements Serializable {
+    private static final long serialVersionUID = 1L;
+
+    @Id
+    private Long id;
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "document_id")
+    // CHECKSTYLE:OFF: MemberName
+    private Document Document;
+    // CHECKSTYLE:ON: MemberName
+
+    public DocumentHolder() {
+    }
+
+    public Long getId() {
+        return id;
+    }
+
+    public void setId(Long id) {
+        this.id = id;
+    }
+
+    public Document getDocument() {
+        return Document;
+    }
+
+    public void setDocument(Document document) {
+        this.Document = document;
+    }
+
+}

--- a/core/testsuite/src/main/resources/META-INF/persistence.xml
+++ b/core/testsuite/src/main/resources/META-INF/persistence.xml
@@ -25,6 +25,7 @@
         <class>com.blazebit.persistence.testsuite.entity.DocumentForOneToOneJoinTable</class>
         <class>com.blazebit.persistence.testsuite.entity.DocumentForOneToOne</class>
         <class>com.blazebit.persistence.testsuite.entity.DocumentForSimpleOneToOne</class>
+        <class>com.blazebit.persistence.testsuite.entity.DocumentHolder</class>
         <class>com.blazebit.persistence.testsuite.entity.DocumentInfo</class>
         <class>com.blazebit.persistence.testsuite.entity.DocumentInfoSimple</class>
         <class>com.blazebit.persistence.testsuite.entity.DocumentNodeCTE</class>

--- a/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/metamodel/MetamodelBuildingContextImpl.java
+++ b/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/metamodel/MetamodelBuildingContextImpl.java
@@ -19,6 +19,7 @@ package com.blazebit.persistence.view.impl.metamodel;
 import com.blazebit.persistence.parser.AliasReplacementVisitor;
 import com.blazebit.persistence.parser.EntityMetamodel;
 import com.blazebit.persistence.parser.expression.AbstractCachingExpressionFactory;
+import com.blazebit.persistence.parser.expression.EntityLiteral;
 import com.blazebit.persistence.parser.expression.Expression;
 import com.blazebit.persistence.parser.expression.ExpressionFactory;
 import com.blazebit.persistence.parser.expression.MacroConfiguration;
@@ -54,6 +55,7 @@ import com.blazebit.persistence.view.spi.type.BasicUserType;
 import com.blazebit.persistence.view.spi.type.TypeConverter;
 import com.blazebit.reflection.ReflectionUtils;
 
+import javax.persistence.metamodel.Attribute;
 import javax.persistence.metamodel.ManagedType;
 import java.lang.annotation.Annotation;
 import java.util.ArrayList;
@@ -346,6 +348,15 @@ public class MetamodelBuildingContextImpl implements MetamodelBuildingContext {
             ));
         }
         Expression simpleExpression = typeValidationExpressionFactory.createSimpleExpression(expression, false, false, true);
+        if (simpleExpression instanceof EntityLiteral) {
+            // This is a special case where the mapping i.e. attribute name matches an entity name
+            try {
+                Attribute<?, ?> attribute = managedType.getAttribute(expression);
+                return Collections.<ScalarTargetResolvingExpressionVisitor.TargetType>singletonList(new ScalarTargetResolvingExpressionVisitor.TargetTypeImpl(false, attribute, attribute.getJavaType(), null, attribute.getJavaType()));
+            } catch (IllegalArgumentException ex) {
+                // Apparently it's not an attribute, so let it run through
+            }
+        }
         ScalarTargetResolvingExpressionVisitor visitor = new ScalarTargetResolvingExpressionVisitor(managedType, entityMetamodel, jpqlFunctions, rootTypes);
         simpleExpression.accept(visitor);
         return visitor.getPossibleTargetTypes();

--- a/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/EntityViewUpdaterImpl.java
+++ b/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/EntityViewUpdaterImpl.java
@@ -1372,19 +1372,17 @@ public class EntityViewUpdaterImpl implements EntityViewUpdater {
         if (!attribute.isOrphanRemoval()) {
             return null;
         }
-
-        if (elementDescriptor.isSubview()) {
-            return new ViewCollectionRemoveListener(elementDescriptor.getLoadOnlyViewToEntityMapper());
-        } else {
-            return EntityCollectionRemoveListener.INSTANCE;
-        }
+        return createCollectionRemoveListener(elementDescriptor);
     }
 
     private static CollectionRemoveListener createCascadeDeleteListener(AbstractMethodAttribute<?, ?> attribute, TypeDescriptor elementDescriptor) {
         if (!attribute.isDeleteCascaded()) {
             return null;
         }
+        return createCollectionRemoveListener(elementDescriptor);
+    }
 
+    private static CollectionRemoveListener createCollectionRemoveListener(TypeDescriptor elementDescriptor) {
         if (elementDescriptor.isSubview()) {
             return new ViewCollectionRemoveListener(elementDescriptor.getLoadOnlyViewToEntityMapper());
         } else {

--- a/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/flush/InverseFlusher.java
+++ b/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/flush/InverseFlusher.java
@@ -22,6 +22,7 @@ import com.blazebit.persistence.parser.EntityMetamodel;
 import com.blazebit.persistence.parser.util.JpaMetamodelUtils;
 import com.blazebit.persistence.spi.ExtendedAttribute;
 import com.blazebit.persistence.spi.ExtendedManagedType;
+import com.blazebit.persistence.view.EntityViewSetting;
 import com.blazebit.persistence.view.OptimisticLockException;
 import com.blazebit.persistence.view.impl.EntityViewManagerImpl;
 import com.blazebit.persistence.view.impl.accessor.Accessors;
@@ -46,6 +47,7 @@ import com.blazebit.persistence.view.impl.type.EntityBasicUserType;
 import com.blazebit.persistence.view.impl.update.EntityViewUpdaterImpl;
 import com.blazebit.persistence.view.impl.update.UpdateContext;
 import com.blazebit.persistence.view.metamodel.ManagedViewType;
+import com.blazebit.persistence.view.metamodel.MethodAttribute;
 import com.blazebit.persistence.view.metamodel.PluralAttribute;
 import com.blazebit.persistence.view.metamodel.SingularAttribute;
 import com.blazebit.persistence.view.metamodel.Type;
@@ -73,6 +75,7 @@ public final class InverseFlusher<E> {
     private final String attributeName;
     private final String parentIdAttributeName;
     private final String childIdAttributeName;
+    private final Class<?> childIdViewClass;
     private final UnmappedAttributeCascadeDeleter deleter;
     // Maps the parent view object to an entity via means of em.getReference
     private final ViewToEntityMapper parentReferenceViewToEntityMapper;
@@ -93,7 +96,7 @@ public final class InverseFlusher<E> {
     private final Mapper<E, Object> parentEntityOnChildEntityMapper;
     private final InverseEntityToEntityMapper childEntityToEntityMapper;
 
-    public InverseFlusher(Class<?> parentEntityClass, String attributeName, String parentIdAttributeName, String childIdAttributeName, UnmappedAttributeCascadeDeleter deleter,
+    public InverseFlusher(Class<?> parentEntityClass, String attributeName, String parentIdAttributeName, String childIdAttributeName, Class<?> childIdViewClass, UnmappedAttributeCascadeDeleter deleter,
                           ViewToEntityMapper parentReferenceViewToEntityMapper, DirtyAttributeFlusher<?, E, Object> parentReferenceAttributeFlusher,
                           Mapper<E, Object> parentEntityOnChildViewMapper, TargetViewClassBasedInverseViewToEntityMapper childViewToEntityMapper, ViewToEntityMapper childReferenceViewToEntityMapper,
                           Mapper<E, Object> parentEntityOnChildEntityMapper, InverseEntityToEntityMapper childEntityToEntityMapper) {
@@ -101,6 +104,7 @@ public final class InverseFlusher<E> {
         this.attributeName = attributeName;
         this.parentIdAttributeName = parentIdAttributeName;
         this.childIdAttributeName = childIdAttributeName;
+        this.childIdViewClass = childIdViewClass;
         this.deleter = deleter;
         this.parentReferenceViewToEntityMapper = parentReferenceViewToEntityMapper;
         this.parentReferenceAttributeFlusher = parentReferenceAttributeFlusher;
@@ -283,6 +287,7 @@ public final class InverseFlusher<E> {
             UnmappedAttributeCascadeDeleter deleter = null;
             String parentIdAttributeName = null;
             String childIdAttributeName = null;
+            Class<?> childIdViewClass = null;
             // Only construct when orphanRemoval or delete cascading is enabled, orphanRemoval implies delete cascading
             if (attribute.isDeleteCascaded()) {
                 EntityMetamodel entityMetamodel = evm.getMetamodel().getEntityMetamodel();
@@ -298,6 +303,13 @@ public final class InverseFlusher<E> {
                     } else {
                         ExtendedAttribute extendedAttribute = elementManagedType.getAttribute(mapping);
                         if (childTypeDescriptor.isSubview()) {
+                            if (elementType instanceof ViewType<?>) {
+                                MethodAttribute<?, ?> idAttribute = ((ViewType<?>) elementType).getIdAttribute();
+                                if (idAttribute.isSubview()) {
+                                    // in this case, we need to fetch the id as view as the deleter expects it this way
+                                    childIdViewClass = idAttribute.getJavaType();
+                                }
+                            }
                             deleter = new ViewTypeCascadeDeleter(childTypeDescriptor.getViewToEntityMapper());
                         } else if (childTypeDescriptor.isJpaEntity()) {
                             deleter = new UnmappedBasicAttributeCascadeDeleter(evm, mapping, extendedAttribute, mapping + "." + parentIdAttributeName, false);
@@ -343,6 +355,7 @@ public final class InverseFlusher<E> {
                     attribute.getMapping(),
                     parentIdAttributeName,
                     childIdAttributeName,
+                    childIdViewClass,
                     deleter,
                     parentReferenceViewToEntityMapper,
                     parentReferenceAttributeFlusher,
@@ -383,11 +396,15 @@ public final class InverseFlusher<E> {
 
     public List<PostFlushDeleter> removeByOwnerId(UpdateContext context, Object ownerId) {
         EntityViewManagerImpl evm = context.getEntityViewManager();
-        List<Object> elementIds = (List<Object>) evm.getCriteriaBuilderFactory().create(context.getEntityManager(), parentEntityClass, "e")
-                .where(parentIdAttributeName).eq(ownerId)
-                .where("e." + attributeName + "." + childIdAttributeName).isNotNull()
-                .select("e." + attributeName + "." + childIdAttributeName)
-                .getResultList();
+        CriteriaBuilder<Object> cb = (CriteriaBuilder<Object>) evm.getCriteriaBuilderFactory().create(context.getEntityManager(), parentEntityClass, "e")
+            .where(parentIdAttributeName).eq(ownerId)
+            .where("e." + attributeName + "." + childIdAttributeName).isNotNull();
+        if (childIdViewClass == null) {
+            cb.select("e." + attributeName + "." + childIdAttributeName);
+        } else {
+            evm.applySetting(EntityViewSetting.create(childIdViewClass), cb, "e." + attributeName + "." + childIdAttributeName);
+        }
+        List<Object> elementIds = cb.getResultList();
         if (!elementIds.isEmpty()) {
             // We must always delete this, otherwise we might get a constraint violation because of the cascading delete
             removeByOwnerIdOnly(context, ownerId);

--- a/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/flush/ViewCollectionRemoveListener.java
+++ b/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/flush/ViewCollectionRemoveListener.java
@@ -35,7 +35,8 @@ public class ViewCollectionRemoveListener implements CollectionRemoveListener {
 
     @Override
     public void onEntityCollectionRemove(UpdateContext context, Object element) {
-        viewToEntityMapper.removeById(context, viewToEntityMapper.getEntityIdAccessor().getValue(element));
+        Object viewId = ((CompositeAttributeFlusher) viewToEntityMapper.getFullGraphNode()).createViewIdByEntityId(viewToEntityMapper.getEntityIdAccessor().getValue(element));
+        viewToEntityMapper.removeById(context, viewId);
     }
 
     @Override

--- a/entity-view/testsuite/src/test/java/com/blazebit/persistence/view/testsuite/basic/AttributeMatchesEntityNameTest.java
+++ b/entity-view/testsuite/src/test/java/com/blazebit/persistence/view/testsuite/basic/AttributeMatchesEntityNameTest.java
@@ -1,0 +1,70 @@
+/*
+ * Copyright 2014 - 2021 Blazebit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.blazebit.persistence.view.testsuite.basic;
+
+import com.blazebit.persistence.testsuite.entity.Document;
+import com.blazebit.persistence.testsuite.entity.DocumentHolder;
+import com.blazebit.persistence.view.EntityView;
+import com.blazebit.persistence.view.EntityViewManager;
+import com.blazebit.persistence.view.IdMapping;
+import com.blazebit.persistence.view.Mapping;
+import com.blazebit.persistence.view.testsuite.AbstractEntityViewTest;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ *
+ * @author Christian Beikov
+ * @since 1.6.0
+ */
+public class AttributeMatchesEntityNameTest extends AbstractEntityViewTest {
+
+    protected EntityViewManager evm;
+
+    @Override
+    protected Class<?>[] getEntityClasses() {
+        return concat(super.getEntityClasses(), DocumentHolder.class);
+    }
+
+    @Before
+    public void initEvm() {
+        evm = build(
+            DocumentHolderView.class,
+            DocumentIdView.class
+        );
+    }
+
+    @EntityView(DocumentHolder.class)
+    public interface DocumentHolderView {
+        @IdMapping
+        Long getId();
+        @Mapping("Document")
+        DocumentIdView getDocument();
+    }
+
+    @EntityView(Document.class)
+    public interface DocumentIdView {
+        @IdMapping
+        Long getId();
+    }
+
+    // Test for https://github.com/Blazebit/blaze-persistence/issues/1294
+    @Test
+    public void testFetch() {
+        evm.find(em, DocumentHolderView.class, 1L);
+    }
+}

--- a/entity-view/testsuite/src/test/java/com/blazebit/persistence/view/testsuite/update/subview/inverse/embedded/complex/EntityViewUpdateSubviewInverseEmbeddedComplexTest.java
+++ b/entity-view/testsuite/src/test/java/com/blazebit/persistence/view/testsuite/update/subview/inverse/embedded/complex/EntityViewUpdateSubviewInverseEmbeddedComplexTest.java
@@ -213,6 +213,29 @@ public class EntityViewUpdateSubviewInverseEmbeddedComplexTest extends AbstractE
         Assert.assertEquals(0, legacyOrder.getPositions().size());
     }
 
+    // Test for https://github.com/Blazebit/blaze-persistence/issues/1299
+    @Test
+    public void testRemoveById() {
+        // Given
+        UpdatableLegacyOrderView newOrder = evm.create(UpdatableLegacyOrderView.class);
+        UpdatableLegacyOrderPositionView position = evm.create(UpdatableLegacyOrderPositionView.class);
+        position.getId().setPositionId(0);
+        position.setArticleNumber("123");
+        newOrder.getPositions().add(position);
+        update(newOrder);
+
+        // When
+        em.clear();
+        evm.remove(em, UpdatableLegacyOrderView.class, newOrder.getId());
+        em.flush();
+        em.clear();
+
+        // Then
+        LegacyOrder legacyOrder = em.find(LegacyOrder.class, newOrder.getId());
+        Assert.assertNull(legacyOrder);
+
+    }
+
     @Test
     public void testAddElementCreateViewToCollection() {
         // Given

